File: src/main/buttons/Button.java
package main.buttons;

import ddf.minim.AudioPlayer;
import main.Entity;
import processing.core.PApplet;
import processing.core.PImage;

public class Button extends Entity {

    public static AudioPlayer click;
    private static final int ALPHA = 80;
    private static PImage error;

    private float x;
    private float y;

    private int width;
    private int height;

    private PImage image;
    protected PImage tintedImage;
    boolean tintOverlay = true;

    private boolean disabled = false;

    protected Button(int width, int height, String path) {
        this(width, height, main.loadImage(path));
    }

    Button(int width, int height, PImage image) {
        this(width, height);
        this.image = image;
        this.image.resize(width, height);
        if (this.image == null) {
            this.image = error.copy();
            this.image.resize(this.width, this.height);
        }
    }

    Button(int width, int height) {
        this.width = width;
        this.height = height;
    }

    public static void setErrorImage() {
        error = main.loadImage("resources/images/error.png");
    }

    void setImage(PImage image) {
        this.image = image;
        if (this.image == null) {
            this.image = error.copy();
            this.image.resize(this.width, this.height);
        }
    }

    //using getImage() here so it's easier to implement ToggleButton
    public void draw(float x, float y) {
        this.x = x;
        this.y = y;

        this.drawButton();

        if (this.inBounds()) {
            if (this.tintOverlay && this.getImage() != null) {
                this.getTintedImage();
                if (this.disabled) {
                    main.tint(255,0,0,ALPHA);
                }
                main.image(this.tintedImage, this.getX(), this.getY());
                main.tint(255);
            } else {
                main.noStroke();
                if (this.disabled) {
                    main.fill(255, 0, 0, ALPHA);
                } else {
                    main.fill(255, ALPHA);
                }
                main.rect(x, y, this.width, this.height);
            }
        }

        this.postDraw();
    }

    protected void getTintedImage() {
        if (this.tintedImage == null) {
            this.tintedImage = brightenImage(this.image);
        }
    }

    protected static PImage brightenImage(PImage image) {
        PImage tintedImage = image.copy();
        for (int i = 0;i < tintedImage.pixels.length;i++) {
            float r = main.red(tintedImage.pixels[i]);
            float g = main.green(tintedImage.pixels[i]);
            float b = main.blue(tintedImage.pixels[i]);
            float a = main.alpha(tintedImage.pixels[i]);
            r = PApplet.constrain(r + ALPHA - 30, 0, 255);
            g = PApplet.constrain(g + ALPHA - 30, 0, 255);
            b = PApplet.constrain(b + ALPHA - 30, 0, 255);
            tintedImage.pixels[i] = main.color(r, g, b, a);
        }
        return tintedImage;
    }

    protected void postDraw() {

    }

    protected void draw() {
        this.draw(this.x, this.y);
    }

    public void mousePressed() {
        if (this.inBounds() && !this.disabled) {
            this.clicked();
            if (main.getSettings().getBoolean("sound")) {
                click.play();
                click.rewind();
            }
        }
    }

    public void clicked() {

    }

    //using getImage() here so it's easier to implement ToggleButton
    protected void drawButton() {
        main.image(this.getImage(), this.x, this.y);
    }

    protected boolean inBounds() {
        return main.mouseX >= this.x && main.mouseX < this.x + this.width && main.mouseY >= this.y && main.mouseY < this.y + this.height;
    }

    public PImage getImage() {
        return this.image;
    }

    protected float getX() {
        return this.x;
    }

    protected float getY() {
        return this.y;
    }

    void incrementX(float x) {
        this.x += x;
    }

    void incrementY(float y) {
        this.y += y;
    }

    void setX(float x) {
        this.x = x;
    }

    void setY(float y) {
        this.y = y;
    }

    public int getWidth() {
        return this.width;
    }

    public int getHeight() {
        return this.height;
    }

    public void setDisabled(boolean disabled) {
        this.disabled = disabled;
    }

}

File: src/main/buttons/Button.kt
package main.buttons

import ddf.minim.AudioPlayer
import main.Main
import processing.core.PImage
import processing.core.PApplet

open class Button  {
    val width: Int
    val height: Int
    protected var x = 0f
    protected var y = 0f
    var image: PImage
    @JvmField
    protected var tintedImage: PImage? = null
    open val tintOverlay = true
    private var disabled = false

    internal constructor(width: Int, height: Int): this(width, height, null as PImage?)

    protected constructor(width: Int, height: Int, path: String?) : this(
        width,
        height,
        Main.loadImage(path)
    )

    private constructor(width: Int, height: Int, image: PImage?) {
        this.width = width
        this.height = height
        this.image = image ?: error.copy()
        this.image.resize(width, height)
    }

//    fun setImage(image: PImage?) {
//        this.image = image
//        if (this.image == null) {
//            this.image = error.copy()
//            this.image!!.resize(width, height)
//        }
//    }

    //using getImage() here so it's easier to implement ToggleButton
    open fun draw(x: Float, y: Float) {
        this.x = x
        this.y = y
        drawButton()
        if (inBounds()) {
            if (tintOverlay) {
                getTintedImage()
                if (disabled) {
                    Main.tint(255f, 0f, 0f, ALPHA.toFloat())
                }
                Main.image(tintedImage, this.x, this.y)
                Main.tint(255)
            } else {
                Main.noStroke()
                if (disabled) {
                    Main.fill(255f, 0f, 0f, ALPHA.toFloat())
                } else {
                    Main.fill(255, ALPHA.toFloat())
                }
                Main.rect(x, y, width.toFloat(), height.toFloat())
            }
        }
        postDraw()
    }

    protected open fun getTintedImage() {
        if (tintedImage == null) {
            tintedImage = brightenImage(image)
        }
    }

    protected open fun postDraw() {}

    protected open fun draw() {
        this.draw(x, y)
    }

    open fun mousePressed() {
        if (inBounds() && !disabled) {
            clicked()
            if (Main.settings.getBoolean("sound")) {
                click.play()
                click.rewind()
            }
        }
    }

    open fun clicked() {}

    //using getImage() here so it's easier to implement ToggleButton
    protected open fun drawButton() {
        Main.image(image, x, y)
    }

    protected open fun inBounds(): Boolean {
        return Main.mouseX >= x && Main.mouseX < x + width && Main.mouseY >= y && Main.mouseY < y + height
    }

    fun incrementX(x: Float) {
        this.x += x
    }

    fun incrementY(y: Float) {
        this.y += y
    }

    fun setDisabled(disabled: Boolean) {
        this.disabled = disabled
    }

    companion object {
        lateinit var click: AudioPlayer
        private const val ALPHA = 80
        lateinit var error: PImage
        fun setErrorImage() {
            error = Main.loadImage("resources/images/error.png")
        }

        @JvmStatic
        protected fun brightenImage(image: PImage): PImage {
            val tintedImage = image.copy()
            for (i in tintedImage.pixels.indices) {
                var r = Main.red(tintedImage.pixels[i])
                var g = Main.green(tintedImage.pixels[i])
                var b = Main.blue(tintedImage.pixels[i])
                val a = Main.alpha(tintedImage.pixels[i])
                r = PApplet.constrain(r + ALPHA - 30, 0f, 255f)
                g = PApplet.constrain(g + ALPHA - 30, 0f, 255f)
                b = PApplet.constrain(b + ALPHA - 30, 0f, 255f)
                tintedImage.pixels[i] = Main.color(r, g, b, a)
            }
            return tintedImage
        }
    }
}